### PR TITLE
Fix Issue #28789: Skipped assets show status 'FAILED' when other assets in the same run fail

### LINKED ISSUE #28789: Skipped assets show status 'FAILED' when other assets in the same run fail
### What's the issue?

(Hi dagster team, I'd appreciate your consideration of this quite frustrating issue which I have been facing for a long time.)

In a nutshell : 

* Two or more partitioned assets in the same job, e.g. A and B
* Run the job
* A fails to materialize
* B is skips materialization

In the `Partitions` tab of asset B, I see a message such as the following :

```
Run 59b3565b failed, and did not materialize this asset.
```

While this satement may be true true, _the run did fail (due to A failing) and B was not materialized (due to B being skipped) _ , it is displaying asset B as failed/red in the sidebar partitions list, and in the partition slice visualization along the top of the view. It would seem to me that this is misrepresenting the status of the asset partition, as B did not _fail_ to materialize, it was skipped.

Image

This is even more apparent if B is an asset that has _previously_ been materialized, but subsequently skips materialization (for example if no upstream data has changed). In other words the previous materialization of B should still represent the most up to date version of the asset. In this case it seems clear that the partition for B should not be represented as FAILED. It should probably be displayed as materialized/green.

**Further information:** 

It looks like the core of the logic is represented partly by these bits of code : 

https://github.com/dagster-io/dagster/blob/master/js_modules/dagster-ui/packages/ui-core/src/assets/AssetPartitionDetail.tsx#L201-L206

https://github.com/dagster-io/dagster/blob/master/python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py#L962-L1048

### What did you expect to happen?

_No response_

### How to reproduce?

_No response_

### Dagster version

dagster, version 1.10.6

### Deployment type

None

### Deployment details

_No response_

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by william-castillo-jr**
## Summary

https://gist.github.com/TiaMarieG/e685ca1ed74092f34b3aa19ce0d1a9e9

## Motivation

CodeDay student open-source contribution.

## How I Tested These Changes

Created test cases to validate the updated logic for various scenarios: 
- A run with both successful and failed materializations.
- A run where assets are skipped due to no upstream changes.
- Edge cases like assets with no prior materializations.

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/assets/AssetPartitionDetail.tsx [hunk 1]
```diff
@@ -45,6 +45,28 @@ import {linkToRunEvent, titleForRun} from '../runs/RunUtils';
...
+      }
+    }
+  }
+`;
+
+export const fetchFailureEvents = async (runId: string, assetKey: AssetKey) => {
+  const {data} = await useQuery(ASSET_FAILURE_EVENTS_QUERY, {
+    variables: {runId, assetKey},
+  });
+
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/assets/AssetPartitionDetail.tsx:63]: `useQuery` is a hook so it can't be used like a normal function here.

**Comment by hellendag** [comment]
@jamiedemaria this looks like it might be related to some of the ongoing work regarding asset failure reasons.

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py [hunk 2]
```diff
@@ -1421,3 +1421,40 @@ class GrapheneAssetNodeOrError(graphene.Union):
...
+        name = "AssetFailureEvents"
+
+class GrapheneAssetFailureEventsQuery(graphene.ObjectType):
+    assetFailureEvents = graphene.Field(
+        GrapheneAssetFailureEvents,
+        runId=graphene.NonNull(graphene.String),
+        assetKey=graphene.NonNull(GrapheneAssetKey),
+    )
+
+    def resolve_assetFailureEvents(self, info, runId, assetKey):
```
Comment (jamiedemaria) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:1445]: we have a similar resolver available [here](https://github.com/dagster-io/dagster/blob/75e073f092cdf6816ff323dc379f149ba47773bb/python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py#L325) that you can query to get the failure events for an asset. I'd prefer to use that resolver since it A) doesn't introduce duplicate query paths to get the same information, and B) requires a limit, so that we don't do an unbounded query against the DB

File: js_modules/dagster-ui/packages/ui-core/src/assets/AssetPartitionDetail.tsx [hunk 3]
```diff
@@ -100,6 +122,14 @@ export const AssetPartitionDetailLoader = (props: {assetKey: AssetKey; partition
     return <AssetPartitionDetailEmpty partitionKey={props.partitionKey} />;
   }
 
+  const assetFailed = latestRunForPartition?.id
+    ? fetchFailureEvents(latestRunForPartition.id, props.assetKey)
```
Comment (jamiedemaria) [inline js_modules/dagster-ui/packages/ui-core/src/assets/AssetPartitionDetail.tsx:126]: unfortunately, I don't think this comparison is doing what you want it to do. The way `ASSET_MATERIALIZATION_FAILURE` works is that it is emitted for every asset that is planned to materialize in a run, but doens't get materialized. So this would include assets that are SKIPPED. Eventually, we want to have an additional value on the `ASSET_MATERIALIZATION_FAILURE` event that distinguishes an asset that didn't materialize because of a step failure and one that got skipped, but that is not in place at the moment.

**Review by jamiedemaria — CHANGES_REQUESTED** [review]
Thanks for contributing! Requesting changes for a couple of reasons:
1. see comment [here](https://github.com/dagster-io/dagster/pull/29935#discussion_r2122043973) - I don't think fetching the failure events and comparing will get you the information you need 
2. `ASSET_MATERIALIZATION_FAILURE` are in development right now, and are not emitted by OSS code yet, so building this UI on top of those events won't be useful until the event is emitted consistently for every run.

**Comment by the4thamigo-uk** [comment]
> @jamiedemaria this looks like it might be related to some of the ongoing work regarding asset failure reasons.

Hi, do you have a reference to this work?

**Comment by jamiedemaria** [comment]
> Hi, do you have a reference to this work?

It's currently a feature we are working on internally, so I don't have code i can share at the moment

**Comment by the4thamigo-uk** [comment]
:(
